tests/integration/test_headers_and_subtotals.py

```diff
@@ -3810,3 +3813,457 @@ def it_provide_residual_test_stats_including_hs(self):
         np.testing.assert_array_equal(slice_.zscores, slice_.residual_test_stats[1])
 
         assert slice_.residual_test_stats.shape == (2, 6, 7)
+
+
+class DescribeIntegrated_SubtotalDifferences(object):
+    """TDD driver(s) for Subtotal Difference insertions."""
+
+    def test_cat_subdiff_count(self, row_subdiff_insertions):
+        strand = Cube(CR.CAT_SUBDIFF, transforms=row_subdiff_insertions).partitions[0]
+        counts = strand.counts
+        # --- Subdiff in position 0 equals id=1 - (id=4 + id=5) ---
+        np.testing.assert_almost_equal(counts[0], counts[1] - (counts[4] + counts[5]))
+
+    def test_cat_subdiff_prop(self, row_subdiff_insertions):
+        strand = Cube(CR.CAT_SUBDIFF, transforms=row_subdiff_insertions).partitions[0]
+        table_proportions = strand.table_proportions
+        # --- Subdiff in position 0 equals id=1 - (id=4 + id=5) ---
+        np.testing.assert_almost_equal(
+            table_proportions[0],
+            table_proportions[1] - (table_proportions[4] + table_proportions[5]),
```

Comment:
Following on to part of the prior comment, I think if you use actual values and avoid repeating the computation in the test that you can combine each set of tests that uses the same fixture. This improves performance and I expect improves the readability as well because it's a lot shorter. So it would be something like:

```python
    def it_computes_subdiff_subtotals_for_various_measures(self, row_subdiff_insertions):
        strand = Cube(CR.CAT_SUBDIFF, transforms=row_subdiff_insertions).partitions[0]

        assert strand.table_proportions[0] == pytest.approx(0.3725221)
        assert strand.row_proportions ...
        assert strand.column_proportions ...
        ...
```

Reply:
This kind of opposes advice Slobodon had, where I originally had numbers, but he argued that it was hard to interpret the meaning of the test because it was just something equals a random number. Do you have advice on how to balance these two concerns?

Reply:
Well, all the integration-test expectations that we have contain "random" numbers, so I don't see how this is different. And duplicating the logic is not testing, it's duplicating the logic; if something went wrong in table-proportions, for example, you could get a passing test (because the logic is the same) for code that produces the wrong answer.

So I'd say using actual values for actuals is the broadly established convention, and with cause. If you wanted to make them somewhat easier to reason about, you can manipulate the fixture values so the actuals come out to some round number or something, like `8 - 2 + 3 = 9` or whatever. But a test is not to be understood in isolation, but paired with the unit (or integrated bit).

Also, I'm not sure how important it is to be able to validate actual values by inspection. Either the author got it right or you're in there modifying the test and need to re-reason about the result yourself. The actual value, once written, is just for regression purposes; to make sure the result didn't _change_ since it was validated by the author. I can't remember having ever checked whether the right values appeared in an expectation written by someone else. I think any effort in that regard is better spent in validating the final result, like Joe does.

Reply:
Sounds good, thanks for explaining, I can feel lots of trade-offs here and so trying to get good guide posts.

Does this look good? 
https://github.com/Crunch-io/crunch-cube/blob/363f7dc4aded1ec32073266c373918ecbef3014f/tests/integration/test_headers_and_subtotals.py#L3842-L3902

I worry that it's past some point of too many `assert`s to be meaningful, but can't tell if that just doesn't matter because we're not expecting to have to read them one at a time.

Reply:
I think this looks good. It's true there are a lot of asserts, but this is an integration test, not a unit test. Also, _Slice is a _value_ object (mostly at least) and to me, this test says: "For the case CAT_X_CAT with these transforms, `slice.values == X`, it's just that values is a collection where each member is distinctly named. In any case, I find it an entirely legitimate approach for this kind of test and there are several other examples of this approach throughout the integration tests.

One important characteristic of this approach is that it avoids a lot of little tests that all load the same fixture, but otherwise don't show that they are closely related. Of course there is the performance penalty involved in that as well, and when you multiply that small difference across a number of integration tests, the test runtime can creep up. I run the tests continuously during development such that they run on every save, and keeping the suite runtime in the single-digit seconds keeps the development tempo snappy, for me at least.